### DISCUSSION TITLE
Auth0 Integration with Uno Platform

### DISCUSSION THREAD

**Discussion by vivara-keithpatton** [unoplatform/uno#16565] [Q&A]
Hi, 
I've spent a few hours on this and could not get Auth0 integrated with Uno Platform despite trying a few options. 

Auth0 (Okta) is a very common authz provider and it provides a number of sdks for native applications, but not Uno Platform.
https://auth0.com/docs/quickstart/native#native

The closest I thought was .Net Maui, which led me to the whole business of using the MauiControls library so I could use the Auth0 provided nuget libary for Maui. 
https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Maui/MauiOverview.html?tabs=single-project

I'm working primarily on Android at present so figured out how to add a callback schema on the MainActivity in the head project so that was ok. 

So i made an authz.xaml ContentView in my MauiControls project:

```
<?xml version="1.0" encoding="utf-8" ?>
<ContentView xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="MyApp.MauiControls.AuthZ">
    <VerticalStackLayout>
      <Label
        Text="Hello, World!"
        SemanticProperties.HeadingLevel="Level1"
        FontSize="32"
        HorizontalOptions="Center"
        x:Name="HelloLabel"/>
      <Label
        Text=""
        HorizontalOptions="Center"
        x:Name="ErrorLabel"/>
      <Label 
        Text="Auth0 Here!"
        VerticalOptions="Center" 
        HorizontalOptions="Center" />
      <Button
        x:Name="LoginBtn"
        Text="Login"
        SemanticProperties.Hint="Login"
        Clicked="OnLoginClicked"
        HorizontalOptions="Center" />
      <Button
        x:Name="LogoutBtn"
        Text="Logout"
        SemanticProperties.Hint="Logout"
        Clicked="OnLogoutClicked"
        HorizontalOptions="Center"
        IsVisible="false" />
  </VerticalStackLayout>
</ContentView>

```

which had some corresponding code behind code provided by Auth- to login and logout. 

```
using Auth0.OidcClient;
using IdentityModel.OidcClient.Browser;
using IdentityModel.OidcClient;

namespace MyApp.MauiControls;

public partial class AuthZ : ContentView
{

    Auth0Client client = new Auth0Client(new Auth0ClientOptions
    {
        Domain = "***",
        ClientId = "***",
        RedirectUri = "myapp://callback",
        PostLogoutRedirectUri = "myapp://callback",
        Scope = "openid profile email"
    });

    public AuthZ()
	{
		InitializeComponent();
	}

    private async void OnLoginClicked(object sender, EventArgs e)
    {
        var extraParameters = new Dictionary<string, string>();
        var audience = ""; // FILL WITH AUDIENCE AS NEEDED

        if (!string.IsNullOrEmpty(audience))
            extraParameters.Add("audience", audience);

        var result = await client.LoginAsync(extraParameters);

        DisplayResult(result);
    }

    private async void OnLogoutClicked(object sender, EventArgs e)
    {
        BrowserResultType browserResult = await client.LogoutAsync();

        if (browserResult != BrowserResultType.Success)
        {
            ErrorLabel.Text = browserResult.ToString();
            return;
        }

        LogoutBtn.IsVisible = false;
        LoginBtn.IsVisible = true;

        HelloLabel.Text = $"Hello, World!";
        ErrorLabel.Text = "";
    }

    private void DisplayResult(LoginResult loginResult)
    {
        if (loginResult.IsError)
        {
            ErrorLabel.Text = loginResult.Error;
            return;
        }

        LogoutBtn.IsVisible = true;
        LoginBtn.IsVisible = false;


        HelloLabel.Text = $"Hello, {loginResult.User.Identity.Name}";
        ErrorLabel.Text = "";
    }
}

```

When I click the Login button, i can see Chrome browser attempt to open, then the entire emulator crashes!!?

What's really putting me off using Uno Platform from this is the amount of time I'm going to need to spend trying to integrate libraries that would just work within .Net Maui that most providers are now supporting with sdks. 

Now, i can also see that Uno Platform has support for authentication but the custom and oidc routes (which may work with Auth0 with some fiddling) appear to require the client secret on the app side, which I won't be doing. 

https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Authentication/AuthenticationOverview.html

Has anyone else successfully integrated an authz provider like Auth0 into Uno Platform, and if so could i get some pointers as to what i'm doing wrong here, thanks

**Comment by dsx1986** [comment]
same issue here. I set up OidcAuthentication in appsettings.json file, and setup auth0 credentials, I got False on macos due to storage issue and app crashed: https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Storage/HowTo-RequiredEntitlements.html

I tried web, I also got False
